evm/spec/framework.tex

```diff
@@ -33,5 +33,119 @@ \subsection{Field selection}
+
+To prove this, the first step is to only select the rows of interest in $S_1$ and $S_2$, and filter out the rest. Let $f_1$ be the filter for $S_1$ and $f_2$ the filter for $S_2$. $f_1$ and $f_2$ are constrained to be in $\{0, 1\}$. $f_1 = 1$ (resp. $f_2 = 1$) whenever the row at hand carries out $Op$ in $S_1$ (resp. in $S_2$), and 0 otherwise. Let also $(\alpha, \beta)$ be two challenges.
+
+The idea is to create subtables $S_1'$ and $S_2'$ of $S_1$ and $S_2$ respectively, such that $f^1 = 1$ and $f^2 = 1$ for all their rows. The columns in the subtables are limited to the ones whose values must be identical (the inputs and outputs of $Op$ in our example).
+
+Let $\{c^{1, i}\}_{i=1}^m$ be the columns in $S_1'$ an $\{c^{2,i}\}_{i=1}^m$ be the columns in $S_2'$.
+
+The prover defines a ``running product'' $Z$ for $S_1'$ such that:
+\begin{gather*}
+  Z^{S_1}_{n-1} = 1 \\
+  Z^{S_1}_{i+1} = Z^{S_1}_i \cdot [f^1_i \cdot \Big(\sum_{j=0}^{m-1} \alpha^j \cdot c^{1, j}_i + \beta \Big) + (1 - f^1_i)]
+\end{gather*}
+Similarly, the prover constructs a running product $Z^{S_2}$for $S_2$. Note that $Z$ is computed ``upside down'': we start with $Z_{n-1} = 1$ and the final product is in $Z_1$. 
+
+On top of the constraints to check that the running products were correctly constructed, the verifier checks that $Z^{S_1}_1 = Z^{S_2}_1$.
+This ensures that the columns in $S_1'$ and the columns in $S_2'$ are permutations of each other.
+
+To sum up, for each STARK $S$, the prover:
+\begin{enumerate}
+  \item constructs a running product $P_i^l$ for each table looking into $S$ (called looking products here),
+  \item constructs a running product $P^S$ for $S$ (called looked product here),
+  \item sends the final value for each running product $P_{i, j}^l$ and $P^S_1$ to the verifier,
+  \item sends a commitment to $P_i^l$  and $P^S$ to the verifier.
+\end{enumerate}
+Then, for each STARK $S$, the verifier:
+\begin{enumerate}
+  \item computes the product $P = \prod_i P_{i, l}^l$,
+  \item checks that $P = P^S_1$,
+  \item checks that each $P_i^l$  and $P^S$ was correctly constructed.
+\end{enumerate}
+
+\subsection{Range-checks}
+\label{rc}
+In most cases, tables deal with U256 words, split into 32-bit limbs (to avoid overflowing the field). To prevent a malicious prover from cheating, it is crucial to range-check those limbs. 
+\subsubsection{What to range-check?}
+One can note that every element that ever appears on the stack has been pushed. Therefore, enforcing a range-check on pushed elements is enough to range-check all elements on the stack. Similarly, all elements in memory must have been written prior, and therefore it is enough to range-check memory writes. However, range-checking the PUSH and MSTORE opcodes is not sufficient.
+\begin{enumerate}
+  \item Pushes and memory writes are range-checked in ``BytePackingStark''.
```

Comment:
Only MSTORE_32BYTES is rangechecked by Bytepacking, MSTORE_GENERAL writes aren't (but have been rangechecked by a previous push).